evennia/contrib/building_menu.py

```diff
@@ -0,0 +1,980 @@
+            menu (BuildingMenu, optional): the parent building menu.
+            on_enter (callable, optional): a callable to call when the
+                    caller enters into the choice.
+            on_nomatch (callable, optional): a callable to call when no
+                    match is entered in the choice.
+            on_leave (callable, optional): a callable to call when the caller
+                    leaves the choice.
+            caller (Account or Object, optional): the caller.
+            obj (Object, optional): the object to edit.
+
+        """
+        self.title = title
+        self.key = key
+        self.aliases = aliases
+        self.attr = attr
+        self.text = text
+        self.glance = glance
+        self.on_enter = on_enter
+        self.on_nomatch = on_nomatch
+        self.on_leave = on_leave
+        self.menu = menu
+        self.caller = caller
+        self.obj = obj
+
+    def __repr__(self):
+        return "<Choice (title={}, key={})>".format(self.title, self.key)
+
+    @property
+    def keys(self):
+        """Return a tuple of keys separated by `sep_keys`."""
+        return tuple(self.key.split(self.menu.sep_keys))
+
+    def format_text(self):
+        """Format the choice text and return it, or an empty string."""
+        text = ""
+        if self.text:
+            text = _call_or_get(self.text, menu=self.menu, choice=self, string="", caller=self.caller, obj=self.obj)
+            text = dedent(text.strip("\n"))
+            text = text.format(obj=self.obj, caller=self.caller)
+
+        return text
+
+    def enter(self, string):
+        """Called when the user opens the choice.
+
+        Args:
+            string (str): the entered string.
+
+        """
+        if self.on_enter:
+            _call_or_get(self.on_enter, menu=self.menu, choice=self, string=string, caller=self.caller, obj=self.obj)
+
+    def nomatch(self, string):
+        """Called when the user entered something in the choice.
+
+        Args:
+            string (str): the entered string.
+
+        Returns:
+            to_display (bool): The return value of `nomatch` if set or
+            `True`.  The rule is that if `no_match` returns `True`,
+            then the choice or menu is displayed.
+
+        """
+        if self.on_nomatch:
+            return _call_or_get(self.on_nomatch, menu=self.menu, choice=self, string=string, caller=self.caller, obj=self.obj)
+
+        return True
+
+    def leave(self, string):
+        """Called when the user closes the choice.
+
+        Args:
+            string (str): the entered string.
+
+        """
+        if self.on_leave:
+            _call_or_get(self.on_leave, menu=self.menu, choice=self, string=string, caller=self.caller, obj=self.obj)
+
+
+class BuildingMenu(object):
+
+    """
+    Class allowing to create and set building menus to edit specific objects.
+
+    A building menu is a kind of `EvMenu` designed to edit objects by
```

Comment:
Should change this - it's not a 'kind of `EvMenu`' since it doesn't use `EvMenu` at all. It doesn't work similarly to `EvMenu` either I don't think.